RIPS/rip-9995.md

```diff
@@ -0,0 +1,199 @@
+
+Talking about Full Native Account Abstraction, the fundamental idea any solution has to address is a
+mechanism for a Smart Contract Account to separate its validation and execution code sections.
+
+RIP-7560 is build on the current Ethereum contract structure, and therefore has little choice but to rely
+on using some higher-level abstraction.
+
+In its current form, RIP-7560 transactions use Solidity method selectors in order to achieve this separation.
+
+This, however, is far from ideal as this approach "leaks" the concept from a programming language widely used
+in the EVM into the core design of the Ethereum protocol.
+
+While there is no purely technical reason to disallow it and there are already instances of Solidity code
+being "enshrined" in the Ethereum protocol, e.g. the validator deposit contract, such violation of abstraction levels
+often lead to unnecessary technical debt and are worth avoiding if possible.
+
+Additionally, method selectors provide very weak indication of the contract developer's decision to become a
+participant in Native Account Abstraction transaction flow.
+The chance of accidentally exposing a function with a colliding 4 byte
+method identifier that returns a valid 4 byte magic to indicate approval is pretty low, but a malicious developer
+can easily hide such a function giving it an innocent name, making it hard to spot a Native Account Abstraction entity.
+
+This issue to some extent is also present in ERC-4337.
+
+As an alternative, if Native Account Abstraction is to be implemented in coordination with EIP-3540,
+relying on the concept of "code sections" it introduces is a superior approach.
+
+## Specification
+
+### Validation and PostTransaction code sections
+
+The EOF format describes a separation of 'code' and 'data' sections,
+marked with `kind_code (0x02)` and `kind_data (0x04)` kind markers.
+
+The contract that may have a role in an Account Abstraction transaction, either as a Sender, a Paymaster or a Deployer,
+has to contain a section marked with the following kind marker:
+
+```
+kind_abstraction = 0x05
+```
+
+This section is equivalent to a code section.
+
+Its code can be executed during a regular transaction execution and has no special effects.
+If it is the first code section of a contract, it can act as an entry point during regular transaction execution.
+
+Only a single section of `kind_abstraction` is allowed in a contract.
+This rule is validated during contract creation.
+
+### Execution entry point for Account Abstraction transaction type participant entity (Sender, Paymaster and Deployer)
+
+During a regular contract code execution, its behaviour is defined as follows by EIP-3540:
+```
+Execution starts at the first byte of code section 0, and pc is set to 0
+```
+
+However, if a contract is referenced in an `AA_TX_TYPE` transaction as a Sender, Paymaster or a Deployer,
+execution starts at the first byte of code section with `kind_abstraction` marker, and `pc` is set to 0.
```

Comment:
a cool thing about the validation sections, is that they don't have to check the caller at all: they can only be called during validation.
still, I think we should declare a distinguished `ORIGIN`. The reason: a modular account might want to recognize it is during validation from a sub-module